django/contrib/admin/widgets.py

```diff
@@ -402,3 +404,88 @@ def __init__(self, attrs=None):
         if attrs is not None:
             final_attrs.update(attrs)
         super(AdminCommaSeparatedIntegerFieldWidget, self).__init__(attrs=final_attrs)
+
+
+class AdminForeignKeyWidget(forms.Select):
+    """
+    ForeignKey widget that loads options via AJAX to avoid rendering large Querysets.
+
+    This class is responsible for rendering the necessary
+    data attributes for select2 as well as adding the static
+    form media.
+    """
+
+    url_name = 'admin:fk_json'
+
+    def __init__(self, model_admin, attrs=None, using=None):
+        self.model_admin = model_admin
+        self.db = using
+        super(AdminForeignKeyWidget, self).__init__(attrs)
+
+    def get_url(self):
+        """Return URL from instance or by reversing :attr:`.url_name`."""
+        return reverse(self.url_name)
+
+    def render(self, name, value, attrs=None):
+        if value is None:
+            value = ''
+        final_attrs = self.build_attrs(attrs, name=name)
+        final_attrs['data-field_identifier'] = self.get_field_identifier(name)
+        output = [format_html('<select{}>', flatatt(final_attrs))]
+        options = self.render_options([value])
+        if options:
+            output.append(options)
+        output.append('</select>')
+        return mark_safe('\n'.join(output))
+
+    def build_attrs(self, extra_attrs=None, **kwargs):
+        """Set select2's AJAX attributes."""
+        attrs = super(AdminForeignKeyWidget, self).build_attrs(extra_attrs=extra_attrs, **kwargs)
+        attrs.setdefault('class', '')
+        attrs.setdefault('data-minimum-input-length', 0)
```

Comment:
A `minimumInputLength` of zero is [the default for Select2](https://github.com/select2/select2/blob/120672dce79b6862cb17a966af06e0617a404b58/src/js/select2/defaults.js#L363).